Public/Src/Tools/Execution.Analyzer/Analyzers.Core/XlgDebugger/JPath/JPath.cs

```diff
@@ -0,0 +1,138 @@
+using System.IO;
+using System.Linq;
+using Antlr4.Runtime;
+using Antlr4.Runtime.Misc;
+using BuildXL.Utilities;
+
+[assembly: CLSCompliant(false)]
+
+namespace BuildXL.Execution.Analyzer.JPath
+{ 
+    /// <summary>
+    /// Static helper methods for parsing/evaluating JPath expressions
+    /// </summary>
+    public static class JPath
+    {
+        class JPathListener : JPathBaseListener, IAntlrErrorListener<IToken>
+        {
+            /// <nodoc />
+            public string FirstError { get; private set; }
+
+            /// <nodoc />
+            public bool HasErrors => FirstError != null;
+
+            /// <nodoc />
+            public void SyntaxError(TextWriter output, IRecognizer recognizer, IToken offendingSymbol, int line, int charPositionInLine, string msg, RecognitionException e)
+            {
+                if (FirstError == null)
+                {
+                    FirstError = msg;
+                }
+            }
+
+            private readonly Stack<ParserRuleContext> m_ruleStack = new Stack<ParserRuleContext>();
+
+            public override void EnterEveryRule([NotNull] ParserRuleContext context)
+            {
+                m_ruleStack.Push(context);
+            }
+
+            public override void ExitEveryRule([NotNull] ParserRuleContext context)
+            {
+                m_ruleStack.Pop();
+            }
+        }
+
+        /// <nodoc />
+        public static Possible<Expr> TryParse(string str)
+        {
+            return TryParseInternal(str).Then(TryConvertInternal);
+        }
+
+        /// <nodoc />
+        public static Possible<Evaluator.Result> TryEval(Evaluator evaluator, Expr expr)
+        {
+            Contract.Requires(evaluator != null);
+            Contract.Requires(expr != null);
+
+            try
+            {
+                return evaluator.Eval(expr);
+            }
+            catch (Exception e)
+            {
+                return new Failure<string>(e.ToString());
+            }
+        }
+
+        /// <nodoc />
+        public static Possible<Evaluator.Result> TryEval(Evaluator evaluator, string expr)
+        {
+            return TryParse(expr).Then(e => TryEval(evaluator, e));
+        }
+
+        internal static Possible<string[]> GetCompletions(Evaluator evaluator, string str)
+        {
+            var lexer = new JPathLexer(new AntlrInputStream(str));
+            var parser = new JPathParser(new CommonTokenStream(lexer));
+            var listener = new JPathListener();
+            parser.AddErrorListener(listener);
+            var exprCst = parser.expr();
+
+            if (listener.HasErrors)
+            {
+                return new Failure<string>("Syntex error: " + listener.FirstError);
+            }
+
+            var exprAst = exprCst.Accept(new AstConverter());
+            var value = evaluator.Eval(exprAst);
+
+            return new[] { "hi", exprAst.Print(), value.Value.ToString() };
+        }

```

Comment:
Why are we returning the string "hi" here?

Reply:
Very good question :)

This whole method is unused and should be deleted.

Thanks for spotting.